1. mon_standard.pkl > array code



In [56]:
import pickle

USE_SUBLABEL = False
URL_PER_SITE = 10
TOTAL_URLS   = 950

# Load the pickle file
print("Loading datafile...")
# with open("/content/mon_standard.pkl", 'rb') as fi: # Path to mon_standard.pkl in Colab
#   data = pickle.load(fi)
with open("C:/Users/jain5/Desktop/기계학습/mon_standard.pkl", 'rb') as fi: # Path to mon_standard.pkl in Colab
    data = pickle.load(fi)
print('loading 완료')
    
X1 = [] # Array to store instances (timestamps) - 19,000 instances, e.g., [[0.0, 0.5, 3.4, ...], [0.0, 4.5, ...], [0.0, 1.5, ...], ... [... ,45.8]]
X2 = [] # Array to store instances (direction*size) - size information
y = [] # Array to store the site of each instance - 19,000 instances, e.g., [0, 0, 0, 0, 0, 0, ..., 94, 94, 94, 94, 94]

# Differentiate instances and sites, and store them in the respective x and y arrays
# x array (direction*timestamp), y array (site label)
for i in range(TOTAL_URLS):
    if USE_SUBLABEL:
        label = i
    else:
        label = i // URL_PER_SITE # Calculate which site's URL the current URL being processed belongs to and set that value as the label. Thus, URLs fetched from the same site are labeled identically.
    for sample in data[i]:
        size_seq = []
        time_seq = []
        for c in sample:
            dr = 1 if c > 0 else -1
            time_seq.append(abs(c))
            size_seq.append(dr * 512)
        X1.append(time_seq)
        X2.append(size_seq)
        y.append(label)
size = len(y)

print(f'Total samples: {size}') # Output: 19000


Loading datafile...
loading 완료
Total samples: 19000


```
IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)
```

data 출력은 안됨. head()로 일부만 하는건 되나?

size는 정상적으로 출력되서 load()가 된건 맞는거 같은데?

In [57]:
import pandas as pd

df = pd.DataFrame(data)

df.head(30)

0    \
0   [0.0, -0.14, 0.14, -0.31, 0.31, -0.51, 0.51, 0...   
1   [0.0, -0.13, 0.13, -0.31, 0.77, -1.11, 1.11, 1...   
2   [0.0, -0.11, 0.11, -0.23, 0.97, -1.11, 1.11, 1...   
3   [0.0, -0.27, 0.27, -0.6, 0.6, -0.88, 0.89, 0.8...   
4   [0.0, -0.11, 0.11, -0.36, 0.36, -0.6, 0.6, 0.6...   
5   [0.0, -0.14, 0.14, -0.3, 1.12, -1.29, 1.29, 1....   
6   [0.0, -0.16, 0.16, -0.74, 2.0, -2.2, 2.2, 2.2,...   
7   [0.0, -0.11, 0.11, -0.23, 0.23, -0.35, 0.35, 0...   
8   [0.0, -0.24, 0.24, -0.49, 0.49, -0.81, 0.82, 0...   
9   [0.0, -0.2, 0.2, -0.49, 1.51, -1.8, 1.8, 1.8, ...   
10  [0.0, -0.13, 0.13, -0.31, 0.86, -1.06, 1.06, 1...   
11  [0.0, -0.73, 0.73, -0.89, 1.79, -1.95, 1.95, 1...   
12  [0.0, -0.24, 0.24, -0.57, 2.83, -3.18, 3.18, 3...   
13  [0.0, -0.14, 0.14, -0.29, 1.16, -1.31, 1.31, 1...   
14  [0.0, -0.23, 0.23, -0.38, 1.11, -1.26, 1.26, 1...   
15  [0.0, -0.1, 0.1, -0.23, 2.31, -2.43, 2.44, 2.4...   
16  [0.0, -0.17, 0.17, -0.33, 0.84, -1.01, 1.02, 1...   
17  [0.0, -0.16, 0.16, -0.34, 1.33, -1.52, 1.52, 1...   
18  [0.0, -0.11, 0.11, -0.3, 0.3, -0.47, 0.47, 0.4...   
19  [0.0, -0.21, 0.21, -0.54, 0.96, -1.72, 1.72, 1...   

                                                  1    \
0   [0.0, -0.93, 0.93, -1.11, 1.11, -1.39, 1.39, 1...   
1   [0.0, -0.11, 0.11, -0.23, 1.15, -1.29, 1.29, 1...   
2   [0.0, -0.15, 0.15, -0.33, 1.27, -1.45, 1.45, 1...   
3   [0.0, -0.14, 0.14, -0.33, 1.14, -1.34, 1.34, 1...   
4   [0.0, -0.12, 0.12, -0.25, 0.25, -0.43, 0.43, 0...   
5   [0.0, -0.23, 0.23, -0.47, 0.48, -0.77, 0.77, 0...   
6   [0.0, -0.12, 0.12, -0.54, 0.54, -0.79, 0.79, 0...   
7   [0.0, -0.12, 0.12, -0.25, 0.25, -0.4, 0.4, 0.4...   
8   [0.0, -0.17, 0.17, -0.4, 0.4, -0.66, 0.66, 0.6...   
9   [0.0, -0.19, 0.19, -0.4, 0.57, -0.72, 0.72, 0....   
10  [0.0, -0.14, 0.14, -0.31, 1.36, -1.55, 1.55, 1...   
11  [0.0, -0.12, 0.12, -0.25, 0.25, -0.4, 0.4, 0.4...   
12  [0.0, -0.12, 0.12, -0.24, 1.02, -1.17, 1.17, 1...   
13  [0.0, -0.24, 0.24, -0.48, 2.61, -2.88, 2.88, 2...   
14  [0.0, -0.11, 0.11, -0.28, 0.35, -0.58, 0.58, 0...   
15  [0.0, -0.3, 0.3, -0.51, 0.51, -0.92, 0.92, 0.9...   
16  [0.0, -0.11, 0.11, -0.23, 1.1, -1.22, 1.23, 1....   
17  [0.0, -0.23, 0.23, -0.37, 1.32, -1.47, 1.47, 1...   
18  [0.0, -0.17, 0.18, -0.43, 0.43, -0.69, 0.7, 0....   
19  [0.0, -0.13, 0.13, -0.52, 0.52, -0.8, 0.8, 0.8...   

                                                  2    \
0   [0.0, -0.15, 0.15, -0.28, 1.24, -1.38, 1.38, 1...   
1   [0.0, -0.27, 0.27, -0.7, 0.7, -1.15, 1.15, 1.1...   
2   [0.0, -0.13, 0.13, -0.3, 0.3, -0.48, 0.48, 0.4...   
3   [0.0, -0.1, 0.1, -0.21, 0.21, -0.34, 0.34, 0.3...   
4   [0.0, -0.16, 0.16, -0.32, 1.17, -1.36, 1.36, 1...   
5   [0.0, -0.14, 0.14, -0.33, 0.71, -1.23, 1.23, 1...   
6   [0.0, -0.12, 0.12, -0.24, 0.89, -1.03, 1.03, 1...   
7   [0.0, -0.13, 0.13, -0.29, 1.01, -1.16, 1.16, 1...   
8   [0.0, -0.1, 0.1, -0.29, 1.56, -1.76, 1.76, 1.7...   
9   [0.0, -0.15, 0.15, -0.36, 0.88, -1.06, 1.06, 1...   
10  [0.0, -0.12, 0.12, -0.27, 0.41, -0.56, 0.56, 0...   
11  [0.0, -0.13, 0.13, -0.28, 0.85, -1.02, 1.02, 1...   
12  [0.0, -0.31, 0.31, -0.53, 0.97, -1.15, 1.15, 1...   
13  [0.0, -0.23, 0.23, -0.52, 0.73, -1.23, 1.23, 1...   
14  [0.0, -0.1, 0.1, -0.24, 3.04, -3.19, 3.19, 3.1...   
15  [0.0, -0.25, 0.25, -0.47, 0.47, -0.81, 0.81, 0...   
16  [0.0, -0.13, 0.13, -0.27, 0.27, -0.43, 0.44, 0...   
17  [0.0, -0.13, 0.13, -0.3, 1.01, -1.18, 1.18, 1....   
18  [0.0, -0.31, 0.31, -0.61, 1.93, -2.24, 2.24, 2...   
19  [0.0, -0.13, 0.13, -0.29, 0.74, -0.91, 0.91, 0...   

                                                  3    \
0   [0.0, -0.1, 0.1, -0.29, 1.18, -1.35, 1.35, 1.3...   
1   [0.0, -0.12, 0.12, -0.26, 1.19, -1.36, 1.36, 1...   
2   [0.0, -0.13, 0.13, -0.32, 0.32, -0.58, 0.58, 0...   
3   [0.0, -0.19, 0.19, -0.49, 0.49, -0.86, 0.87, 0...   
4   [0.0, -0.24, 0.24, -0.64, 1.3, -1.82, 1.83, 1....   
5   [0.0, -0.16, 0.16, -0.29, 0.29, -0.45, 0.45, 0...   
6   [0.0, -0.2, 0.2, -0.44, 1.23, -1.6, 1.61, 1.61...   


In [58]:
len(X1) #19000 -> Null 데이터 없음
len(X1[0]) #1421
#len(X1[18999]) # 9925

1421

In [61]:
# 관찰별로 들어오는 송신/수신 패킷의 누적합
incoming_packets_size=[0]*19000# Server->Client
outgoing_packets_size=[0]*19000
i=0
j=0
for i in range(19000):
    for j in range(len(X2[i])):
        if X2[i][j]<0: # X2[0] -> 0번 인덱스에는 1번 웹페이지의 1번째 관찰 결과가 리스트로 저장. 1번 인덱스는 1번 웹페이지의 2번째 관찰 결과...이렇게 20회씩 관찰
            incoming_packets_size[i]+=X2[i][j] # 음수인거..즉 수신 packet size 누적합
        else:
            outgoing_packets_size[i]+=X2[i][j] # 양수...송신 packet size 누적합
len(outgoing_packets_size) #19000

19000

In [62]:
# 전체 누적 합
total_packets_size=[0]*190000

i=0
for i in range(19000):
    total_packets_size[i] = incoming_packets_size[i] + outgoing_packets_size[i]
len(total_packets_size)

190000

In [73]:
# print(X1[0]) #0.0~5.56, 130개
# print((X1[1])) #0.0~11.88, 9927개
# print((X1[2])) #0.0~17.81 ,359개
# 총 패킷 수에서 분율로 나타나는 송신 패킷 수 

In [74]:
len(X2[0])#1421
X2[0][0]

-512

In [75]:
from collections import Counter

print(Counter(y)) # 200개씩 label

Counter({0: 200, 1: 200, 2: 200, 3: 200, 4: 200, 5: 200, 6: 200, 7: 200, 8: 200, 9: 200, 10: 200, 11: 200, 12: 200, 13: 200, 14: 200, 15: 200, 16: 200, 17: 200, 18: 200, 19: 200, 20: 200, 21: 200, 22: 200, 23: 200, 24: 200, 25: 200, 26: 200, 27: 200, 28: 200, 29: 200, 30: 200, 31: 200, 32: 200, 33: 200, 34: 200, 35: 200, 36: 200, 37: 200, 38: 200, 39: 200, 40: 200, 41: 200, 42: 200, 43: 200, 44: 200, 45: 200, 46: 200, 47: 200, 48: 200, 49: 200, 50: 200, 51: 200, 52: 200, 53: 200, 54: 200, 55: 200, 56: 200, 57: 200, 58: 200, 59: 200, 60: 200, 61: 200, 62: 200, 63: 200, 64: 200, 65: 200, 66: 200, 67: 200, 68: 200, 69: 200, 70: 200, 71: 200, 72: 200, 73: 200, 74: 200, 75: 200, 76: 200, 77: 200, 78: 200, 79: 200, 80: 200, 81: 200, 82: 200, 83: 200, 84: 200, 85: 200, 86: 200, 87: 200, 88: 200, 89: 200, 90: 200, 91: 200, 92: 200, 93: 200, 94: 200})


In [76]:
# Total Number of packets, array는 레이블별 평균 총 패킷 수
import numpy as np
N_packets=[]
i=0
for i in range(len(X1)):
    N_packets.append(len(X1[i])) # N_packets는 모든 관찰 횟수
#print(N_packets) # 19000

In [77]:
# 관찰별로 들어오는 송신/수신 패킷의 수
incoming_packets_N=[0]*19000# Server->Client
outgoing_packets_N=[0]*19000
i=0
j=0
for i in range(19000):
    for j in range(len(X2[i])):
        if X2[i][j]<0: # X2[0] -> 0번 인덱스에는 1번 웹페이지의 1번째 관찰 결과가 리스트로 저장. 1번 인덱스는 1번 웹페이지의 2번째 관찰 결과...이렇게 20회씩 관찰
            incoming_packets_N[i]+=1
        else:
            outgoing_packets_N[i]+=1 
len(outgoing_packets_N) #19000

19000

In [78]:
# 송신 패킷 수/전체 패킷수
a=[]
outgoing_fraction=[]
for i in range(19000): 
    a.append(outgoing_packets_N[i]/len(X1[i])) # 일단 fraction 전체(19000) 관찰 결과별로 저장


In [79]:
# 총 패킷 수에서 분율로 나타나는 수신 패킷 수(Number of incoming packets as a fraction of the total number of packets)
b=[]
incoming_fraction=[]
for i in range(19000): 
    b.append(incoming_packets_N[i]/len(X1[i])) # 일단 fraction 전체(19000) 관찰 결과별로 저장


In [80]:
incoming_packets_three=[0]*19000# Server->Client
outgoing_packets_three=[0]*19000
i=0
j=0
for i in range(19000):
    for j in range(30): # 처음 관찰하는 30개의 패킷만 관찰
        if X2[i][j]<0:
            incoming_packets_three[i]+=1
        else:
            outgoing_packets_three[i]+=1 
len(outgoing_packets_three) # 19000

19000

In [81]:
# 추가 feature, 시간당 총 패킷 수 비율
c=[]
time_fraction=[]
for i in range(19000): 
    c.append(N_packets[i]/X1[i][-1]) # 일단 fraction 전체(19000) 관찰 결과별로 저장
len(c)

19000

In [82]:
# burst 확인, 없는거 확인했으니 돌리지 말자..오래걸려...
#X2[1].count([-512, 512])# 0
#i=0
#j=0
#for i in range(19000):
#    for j in range(len(X2[i])):
#        if X2[i][j] not in [-512,512]:
#            print(i,'인덱스', j, '번', X2[i][j])
# burst data 없나봐...그리고 -1024,1024 byte도 없는듯?

index당 20 entris, 950개의 열이므로

총 인스턴스(객체) 수는 950*20 = 1900 == size

각 type은 object
```
1. 1. **mon_standard.pkl**: 이 파일에는 "모니터링"된 웹 사이트의 데이터가 포함되어 있습니다.

   - - 학급수 : 95

   - - 인스턴스 수 : 19,000개 (95개 웹사이트, 비인덱스 페이지 10개의 하위 페이지 각각 20회 관찰)
```

In [85]:
out_std=[]
in_std=[]
i=0
j=0
a1=[]
a2=[]
for i in range(19000):
    for j in range(len(X2[i])):
        if X2[i][j]<0:
            a1.append(j+1) # a1에는 수신(incoming) 패킷 목록 저장
        else:
            a2.append(j+1)
    out_std.append(np.std(a1))
    a1=[]
    in_std.append(np.std(a2))   
    a2=[]
len(out_std)

19000

In [86]:
out_mean=[]
in_mean=[]
i=0
j=0
a1=[]
a2=[]
for i in range(19000):
    for j in range(len(X2[i])):
        if X2[i][j]<0:
            a1.append(j+1) # a1에는 수신(incoming) 패킷 목록 저장
        else:
            a2.append(j+1)
    out_mean.append(np.mean(a1))
    a1=[]
    in_mean.append(np.mean(a2))   
    a2=[]
len(out_mean)

19000

In [87]:
# sum_in + out + total(8)
sum_N=[]
i=0
for i in range(19000):
    sum_N.append(incoming_packets_N[i]+outgoing_packets_N[i]+N_packets[i])
len(sum_N)

19000

In [88]:
import pandas as pd

# index 0~19(20개씩)은 같은 페이지 관찰 결과이다.
my_df=pd.DataFrame(zip(incoming_packets_N, outgoing_packets_N, incoming_packets_three, outgoing_packets_three, a, b,
                      out_std, in_std, out_mean, in_mean, N_packets,c,sum_N, incoming_packets_size, outgoing_packets_size, total_packets_size, y), 
                   columns=['N_in_packets','N_out_packets','in_first_thirty','out_first_thirty', 'out_fraction','in_fraction',
                           'out_std','in_std', 'out_mean', 'in_mean','N_packets', 'time_fraction','sum_N','incoming_packets_size',
                            'outgoing_packets_size','total_packets_size','y'])

my_df

N_in_packets  N_out_packets  in_first_thirty  out_first_thirty  \
0              1300            121               21                 9   
1               438             80               22                 8   
2              1240            118               23                 7   
3              1324            122               21                 9   
4              1291            115               22                 8   
...             ...            ...              ...               ...   
18995          8815            619               23                 7   
18996          9404            552               23                 7   
18997          9373            579               22                 8   
18998          9236            690               23                 7   
18999          9168            757               23                 7   

       out_fraction  in_fraction      out_std       in_std     out_mean  \
0          0.085151     0.914849   398.486335   515.483953   705.106154   
1          0.154440     0.845560   150.591146   139.231951   265.406393   
2          0.086892     0.913108   381.878642   472.735508   669.259677   
3          0.084371     0.915629   406.205288   513.916038   714.503021   
4          0.081792     0.918208   395.028889   503.993490   695.740511   
...             ...          ...          ...          ...          ...   
18995      0.065614     0.934386  2698.463332  3053.116218  4708.505615   
18996      0.055444     0.944556  2863.806259  3010.091146  5004.064653   
18997      0.058179     0.941821  2857.611288  3102.381602  4989.437000   
18998      0.069514     0.930486  2844.328152  3116.574388  4939.918796   
18999      0.076272     0.923728  2825.280155  3269.954300  4922.498146   

           in_mean  N_packets  time_fraction  sum_N  incoming_packets_size  \
0       774.322314       1421     140.138067   2842                -665600   
1       227.162500        518      50.984252   1036                -224256   
2       787.110169       1358     122.232223   2716                -634880   
3       821.139344       1446     108.233533   2892                -677888   
4       790.608696       1406     132.142857   2812                -660992   
...            ...        ...            ...    ...                    ...   
18995  4845.586430       9434     214.848554  18868               -4513280   
18996  4542.974638       9956     638.205128  19912               -4814848   
18997  4767.072539       9952     666.577361  19904               -4798976   
18998  5279.146377       9926     498.543446  19852               -4728832   
18999  5453.516513       9925     721.293605  19850               -4694016   

       outgoing_packets_size  total_packets_size   y  
0                      61952             -603648   0  
1                      40960             -183296   0  
2                      60416             -574464   0  
3                      62464             -615424   0  
4                      58880             -602112   0  
...                      ...                 ...  ..  
18995                 316928            -4196352  94  
18996                 282624            -4532224  94  
18997                 296448            -4502528  94  
18998                 353280            -4375552  94  
18999                 387584            -4306432  94  

[19000 rows x 17 columns]

In [89]:
# dataset 바탕화면에 저장
import pandas as pd
import os

os.chdir("C:/Users/jain5/Desktop")
my_df.to_csv("dataset.csv")

## --------------------------!여기 아래는 test data! label 없음!---------------------------------------

2. unmon_standard10.pkl > array code

In [90]:
import pickle

TOTAL_URLS = 10000  # total number in the dataset

# Load 10,000 unmon pickle file
print("Loading datafile...")
with open("C:/Users/jain5/Desktop/기계학습/unmon_standard10.pkl", 'rb') as f:  # Path to unmon_standard10.pkl in Colab
    x = pickle.load(f)
print('loading 완료')

size = len(x)
print(f'Total samples: {size}')

X1 = [] # Array to store instances (timestamps) - 10,000 instances, e.g., [[0.0, 0.5, 3.4, ...], [0.0, 4.5, ...], [0.0, 1.5, ...], ... [... ,45.8]]
X2 = [] # Array to store instances (direction*size) - size information

for i in range(TOTAL_URLS):
    size_seq = []
    time_seq = []
    for c in x[i]:
        dr = 1 if c > 0 else -1
        time_seq.append(abs(c))
        size_seq.append(dr * 512) # In the pickle file, there is no size information, so the conversion code is set to multiply by 512 uniformly.
    X1.append(time_seq)
    X2.append(size_seq)

print(len(X1)) # Print the length of X1



Loading datafile...
loading 완료
Total samples: 10000
10000


In [91]:
import pandas as pd

df2 = pd.DataFrame(x)

df2.head(3)

0     1     2     3     4     5     6     7     8     9     ...  9982  \
0   0.0 -0.13  0.13 -0.36  0.83 -1.10  1.10 -1.35 -1.35  2.10  ...   NaN   
1   0.0 -0.21  0.21 -0.39  0.85 -1.36  1.36 -1.58 -1.58  3.13  ...   NaN   
2   0.0 -0.23  0.24 -0.61  1.84 -2.28  2.29 -2.69 -2.69  3.71  ...   NaN   

   9983  9984  9985  9986  9987  9988  9989  9990  9991  
0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
1   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  

[3 rows x 9992 columns]

In [92]:
#df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Columns: 9992 entries, 0 to 9991
dtypes: float64(9992)
memory usage: 762.3 MB


```
2. 2. **unmon_standard10.pkl**: 이 파일에는 "모니터링되지 않은" 웹 사이트의 데이터가 포함되어 있습니다.

   - - 사례 수 : 19,000건
```

10000 RangeIndex....10000개 행...load할때 기본 20분은 잡자.

type은 다 float형이고, 열은 9992개

In [93]:
print(len(X1))
print(len(X2))

10000
10000


In [95]:
# 관찰별로 들어오는 송신/수신 패킷의 누적합
unmon_incoming_packets_size=[0]*10000# Server->Client
unmon_outgoing_packets_size=[0]*10000
i=0
j=0
for i in range(10000):
    for j in range(len(X2[i])):
        if X2[i][j]<0: # X2[0] -> 0번 인덱스에는 1번 웹페이지의 1번째 관찰 결과가 리스트로 저장. 1번 인덱스는 1번 웹페이지의 2번째 관찰 결과...이렇게 20회씩 관찰
            unmon_incoming_packets_size[i]+=X2[i][j] # 음수인거..즉 수신 packet size 누적합
        else:
            unmon_outgoing_packets_size[i]+=X2[i][j] # 양수...송신 packet size 누적합
len(unmon_outgoing_packets_size) #19000

10000

In [97]:
# 전체 누적 합
unmon_total_packets_size=[0]*100000

i=0
for i in range(10000):
    unmon_total_packets_size[i] = unmon_incoming_packets_size[i] + unmon_outgoing_packets_size[i]
len(unmon_total_packets_size)

100000

In [98]:
# 관찰별로 들어오는 송신/수신 패킷의 수
unmon_incoming_packets_N=[0]*10000# Server->Client
unmon_outgoing_packets_N=[0]*10000
i=0
j=0
for i in range(len(X1)):
    for j in range(len(X2[i])):
        if X2[i][j]<0: # X2[0] -> 0번 인덱스에는 1번 웹페이지의 1번째 관찰 결과가 리스트로 저장. 1번 인덱스는 1번 웹페이지의 2번째 관찰 결과...이렇게 20회씩 관찰
            unmon_incoming_packets_N[i]+=1
        else:
            unmon_outgoing_packets_N[i]+=1 
len(unmon_outgoing_packets_N) #10000

10000

In [99]:
unmon_incoming_packets_three=[0]*10000# Server->Client
unmon_outgoing_packets_three=[0]*10000
i=0
j=0
for i in range(10000):
    for j in range(30): # 처음 관찰하는 30개의 패킷만 관찰
        if X2[i][j]<0:
            unmon_incoming_packets_three[i]+=1
        else:
            unmon_outgoing_packets_three[i]+=1 
len(unmon_outgoing_packets_three) # 10000

10000

In [100]:
# 송신 패킷 수/전체 패킷수
a=[]
i=0
unmon_outgoing_fraction=[]
for i in range(10000): 
    a.append(unmon_outgoing_packets_N[i]/len(X1[i])) # 일단 fraction 전체(10000) 관찰 결과별로 저장

In [101]:
# 총 패킷 수에서 분율로 나타나는 수신 패킷 수(Number of incoming packets as a fraction of the total number of packets)
b=[]
i=0
unmon_incoming_fraction=[]
for i in range(10000): 
    b.append(unmon_incoming_packets_N[i]/len(X1[i])) # 일단 fraction 전체(10000) 관찰 결과별로 저장

In [102]:
unmon_out_std=[]
unmon_in_std=[]
i=0
j=0
a1=[]
a2=[]
for i in range(10000):
    for j in range(len(X2[i])):
        if X2[i][j]<0:
            a1.append(j+1) # a1에는 수신(incoming) 패킷 목록 저장
        else:
            a2.append(j+1)
    unmon_out_std.append(np.std(a1))
    a1=[]
    unmon_in_std.append(np.std(a2))   
    a2=[]
len(unmon_out_std)

10000

In [103]:
unmon_out_mean=[]
unmon_in_mean=[]
i=0
j=0
a1=[]
a2=[]
for i in range(10000):
    for j in range(len(X2[i])):
        if X2[i][j]<0:
            a1.append(j+1) # a1에는 수신(incoming) 패킷 목록 저장
        else:
            a2.append(j+1)
    unmon_out_mean.append(np.mean(a1))
    a1=[]
    unmon_in_mean.append(np.mean(a2))   
    a2=[]
len(unmon_out_mean)

10000

In [104]:
# Total Number of packets, array는 레이블별 평균 총 패킷 수
import numpy as np
unmon_N_packets=[]
i=0
for i in range(len(X1)):
    unmon_N_packets.append(len(X1[i])) # unmon_N_packets는 모든 관찰 횟수
print(len(unmon_N_packets)) # 10000

10000


In [105]:
# 추가 feature, 시간당 총 패킷 수 비율
c=[]
unmon_time_fraction=[]
for i in range(10000): 
    c.append(unmon_N_packets[i]/X1[i][-1]) # 일단 fraction 전체(1000) 관찰 결과별로 저장
len(c)

10000

In [106]:
y=[-1]*10000

In [107]:
# sum_in + out + total(8)
unmon_sum_N=[]
i=0
for i in range(10000):
    unmon_sum_N.append(unmon_incoming_packets_N[i]+unmon_outgoing_packets_N[i]+unmon_N_packets[i])
len(unmon_sum_N)

10000

In [108]:
import pandas as pd

# index 0~19(20개씩)은 같은 페이지 관찰 결과이다.
unmon_df=pd.DataFrame(zip(unmon_incoming_packets_N, unmon_outgoing_packets_N, unmon_incoming_packets_three, unmon_outgoing_packets_three, a, b,
                      unmon_out_std, unmon_in_std, unmon_out_mean, unmon_in_mean, unmon_N_packets,c,unmon_sum_N, unmon_incoming_packets_size, 
                          unmon_outgoing_packets_size, unmon_total_packets_size, y), 
                   columns=['unmon_N_in_packets','unmon_N_out_packets','unmon_in_first_thirty','unmon_out_first_thirty', 'unmon_out_fraction',
                            'unmon_in_fraction','unmon_out_std','unmon_in_std', 'unmon_out_mean', 'unmon_in_mean','unmon_N_packets',
                            'unmon_time_fraction','unmon_sum_N', 'unmon_incoming_packets_size', 
                            'unmon_outgoing_packets_size', 'unmon_total_packets_size', 'y'])

unmon_df

unmon_N_in_packets  unmon_N_out_packets  unmon_in_first_thirty  \
0                    101                   29                     17   
1                   9189                  738                     18   
2                    285                   74                     17   
3                   9407                  541                     21   
4                   2215                  293                     27   
...                  ...                  ...                    ...   
9995                4180                  413                     20   
9996                4663                  447                     19   
9997                 302                   59                     19   
9998                 413                   96                     21   
9999                9668                  322                     21   

      unmon_out_first_thirty  unmon_out_fraction  unmon_in_fraction  \
0                         13            0.223077           0.776923   
1                         12            0.074343           0.925657   
2                         13            0.206128           0.793872   
3                          9            0.054383           0.945617   
4                          3            0.116826           0.883174   
...                      ...                 ...                ...   
9995                      10            0.089919           0.910081   
9996                      11            0.087476           0.912524   
9997                      11            0.163435           0.836565   
9998                       9            0.188605           0.811395   
9999                       9            0.032232           0.967768   

      unmon_out_std  unmon_in_std  unmon_out_mean  unmon_in_mean  \
0         36.304696     27.053784       73.465347      37.758621   
1       2845.096286   2885.461055     5053.770595    3846.246612   
2        100.730751    111.145044      185.996491     156.905405   
3       2855.784563   3120.454082     4992.069735    4668.994455   
4        716.374241    778.715376     1257.969752    1228.269625   
...             ...           ...             ...            ...   
9995    1337.407077   1173.380403     2271.961722    2550.414044   
9996    1450.994832   1621.869237     2506.849024    3063.015660   
9997     101.241755    118.245320      181.175497     180.101695   
9998     138.848134    166.667122      242.169492     310.197917   
9999    2876.437267   3076.060409     5007.719797    4628.602484   

      unmon_N_packets  unmon_time_fraction  unmon_sum_N  \
0                 130            23.381295          260   
1                9927           835.606061        19854   
2                 359            20.157215          718   
3                9948          1084.841876        19896   
4                2508            46.799776         5016   
...               ...                  ...          ...   
9995             4593           143.128701         9186   
9996             5110           132.314863        10220   
9997              361            10.334956          722   
9998              509            42.989865         1018   
9999             9990          1038.461538        19980   

      unmon_incoming_packets_size  unmon_outgoing_packets_size  \
0                          -51712                        14848   
1                        -4704768                       377856   
2                         -145920                        37888   
3                        -4816384                       276992   
4                        -1134080                       150016   
...                           ...                          ...   
9995                     -2140160                       211456   
9996                     -2387456                       228864   
9997                      -154624                        30208   
9998                      -211456                        49152   
9999                     -4950016   

In [109]:
# dataset 바탕화면에 저장
import pandas as pd
import os

os.chdir("C:/Users/jain5/Desktop")
unmon_df.to_csv("unmon_dataset.csv")

In [110]:
import pandas as pd
df1 = pd.read_csv(r"C:\Users\jain5\Desktop\dataset.csv")
df2 = pd.read_csv(r"C:\Users\jain5\Desktop\unmon_dataset.csv")

In [111]:
df1['y']=[1]*19000
# label 1로 재할당

In [112]:
# 합치기 위해 column name 변경
df2.columns=['N_in_packets','N_out_packets','in_first_thirty','out_first_thirty', 'out_fraction','in_fraction',
                           'out_std','in_std', 'out_mean', 'in_mean','N_packets', 'time_fraction','sum_N','incoming_packets_size',
                            'outgoing_packets_size','total_packets_size','y']

In [113]:
# df 합치기(가로로)
df=pd.concat([df1, df2])

In [114]:
df

N_in_packets  N_out_packets  in_first_thirty  out_first_thirty  \
0             1300            121               21                 9   
1              438             80               22                 8   
2             1240            118               23                 7   
3             1324            122               21                 9   
4             1291            115               22                 8   
...            ...            ...              ...               ...   
9995          4180            413               20                10   
9996          4663            447               19                11   
9997           302             59               19                11   
9998           413             96               21                 9   
9999          9668            322               21                 9   

      out_fraction  in_fraction      out_std       in_std     out_mean  \
0         0.085151     0.914849   398.486335   515.483953   705.106154   
1         0.154440     0.845560   150.591146   139.231951   265.406393   
2         0.086892     0.913108   381.878642   472.735508   669.259677   
3         0.084371     0.915629   406.205288   513.916038   714.503021   
4         0.081792     0.918208   395.028889   503.993490   695.740511   
...            ...          ...          ...          ...          ...   
9995      0.089919     0.910081  1337.407077  1173.380403  2271.961722   
9996      0.087476     0.912524  1450.994832  1621.869237  2506.849024   
9997      0.163435     0.836565   101.241755   118.245320   181.175497   
9998      0.188605     0.811395   138.848134   166.667122   242.169491   
9999      0.032232     0.967768  2876.437267  3076.060409  5007.719797   

          in_mean  N_packets  time_fraction  sum_N  incoming_packets_size  \
0      774.322314       1421     140.138067   2842                -665600   
1      227.162500        518      50.984252   1036                -224256   
2      787.110169       1358     122.232223   2716                -634880   
3      821.139344       1446     108.233533   2892                -677888   
4      790.608696       1406     132.142857   2812                -660992   
...           ...        ...            ...    ...                    ...   
9995  2550.414044       4593     143.128701   9186               -2140160   
9996  3063.015660       5110     132.314863  10220               -2387456   
9997   180.101695        361      10.334956    722                -154624   
9998   310.197917        509      42.989865   1018                -211456   
9999  4628.602484       9990    1038.461538  19980               -4950016   

      outgoing_packets_size  total_packets_size  y  
0                     61952             -603648  1  
1                     40960             -183296  1  
2                     60416             -574464  1  
3                     62464             -615424  1  
4                     58880             -602112  1  
...                     ...                 ... ..  
9995                 211456            -1928704 -1  
9996                 228864            -2158592 -1  
9997                  30208             -124416 -1  
9998                  49152             -162304 -1  
9999                 164864            -4785152 -1  

[29000 rows x 17 columns]

In [115]:
# dataset 바탕화면에 저장
import pandas as pd
import os

os.chdir("C:/Users/jain5/Desktop")
df.to_csv("open_binary.csv")

In [116]:
import pandas as pd
df1 = pd.read_csv(r"C:\Users\jain5\Desktop\dataset.csv")
df2 = pd.read_csv(r"C:\Users\jain5\Desktop\unmon_dataset.csv")

In [117]:
df2.columns=['N_in_packets','N_out_packets','in_first_thirty','out_first_thirty', 'out_fraction','in_fraction',
                           'out_std','in_std', 'out_mean', 'in_mean','N_packets', 'time_fraction','sum_N','incoming_packets_size',
                            'outgoing_packets_size','total_packets_size','y']

In [118]:
# df 합치기(가로로)
df=pd.concat([df1, df2])

In [119]:
df

N_in_packets  N_out_packets  in_first_thirty  out_first_thirty  \
0             1300            121               21                 9   
1              438             80               22                 8   
2             1240            118               23                 7   
3             1324            122               21                 9   
4             1291            115               22                 8   
...            ...            ...              ...               ...   
9995          4180            413               20                10   
9996          4663            447               19                11   
9997           302             59               19                11   
9998           413             96               21                 9   
9999          9668            322               21                 9   

      out_fraction  in_fraction      out_std       in_std     out_mean  \
0         0.085151     0.914849   398.486335   515.483953   705.106154   
1         0.154440     0.845560   150.591146   139.231951   265.406393   
2         0.086892     0.913108   381.878642   472.735508   669.259677   
3         0.084371     0.915629   406.205288   513.916038   714.503021   
4         0.081792     0.918208   395.028889   503.993490   695.740511   
...            ...          ...          ...          ...          ...   
9995      0.089919     0.910081  1337.407077  1173.380403  2271.961722   
9996      0.087476     0.912524  1450.994832  1621.869237  2506.849024   
9997      0.163435     0.836565   101.241755   118.245320   181.175497   
9998      0.188605     0.811395   138.848134   166.667122   242.169491   
9999      0.032232     0.967768  2876.437267  3076.060409  5007.719797   

          in_mean  N_packets  time_fraction  sum_N  incoming_packets_size  \
0      774.322314       1421     140.138067   2842                -665600   
1      227.162500        518      50.984252   1036                -224256   
2      787.110169       1358     122.232223   2716                -634880   
3      821.139344       1446     108.233533   2892                -677888   
4      790.608696       1406     132.142857   2812                -660992   
...           ...        ...            ...    ...                    ...   
9995  2550.414044       4593     143.128701   9186               -2140160   
9996  3063.015660       5110     132.314863  10220               -2387456   
9997   180.101695        361      10.334956    722                -154624   
9998   310.197917        509      42.989865   1018                -211456   
9999  4628.602484       9990    1038.461538  19980               -4950016   

      outgoing_packets_size  total_packets_size  y  
0                     61952             -603648  0  
1                     40960             -183296  0  
2                     60416             -574464  0  
3                     62464             -615424  0  
4                     58880             -602112  0  
...                     ...                 ... ..  
9995                 211456            -1928704 -1  
9996                 228864            -2158592 -1  
9997                  30208             -124416 -1  
9998                  49152             -162304 -1  
9999                 164864            -4785152 -1  

[29000 rows x 17 columns]

In [120]:
import pandas as pd
import os

os.chdir("C:/Users/jain5/Desktop")
df.to_csv("open_multi.csv")